synfig-core/test/string.cpp

```diff
@@ -0,0 +1,187 @@
+{
+	std::string res = remove_trailing_zeroes("10.0305");
+	ASSERT_EQUAL("10.0305", res)
+}
+
+void
+test_remove_trailing_zeroes_without_trailing_zeroes_and_no_decimal_point_appends_dot_zero()
+{
+	std::string res = remove_trailing_zeroes("102");
+	ASSERT_EQUAL("102.0", res)
+}
+
+void
+test_remove_trailing_zeroes_with_trailing_zeroes_but_no_decimal_point_appends_dot_zero()
+{
+	std::string res = remove_trailing_zeroes("100");
+	ASSERT_EQUAL("100.0", res)
+}
+
+void
+test_remove_trailing_zeroes_without_trailing_zeroes_and_no_decimal_point_does_not_appends_dot_zero_if_requested()
+{
+	std::string res = remove_trailing_zeroes("102", false);
+	ASSERT_EQUAL("102", res)
+}
+
+void
+test_remove_trailing_zeroes_with_trailing_zeroes_but_no_decimal_point_does_not_appends_dot_zero_if_requested()
+{
+	std::string res = remove_trailing_zeroes("100", false);
+	ASSERT_EQUAL("100", res)
+}
+void
+test_remove_trailing_zeroes_with_real_trailing_zeroes_does_remove_them()
+{
+	std::string res = remove_trailing_zeroes("100.3000");
+	ASSERT_EQUAL("100.3", res)
+}
+
+void
+test_remove_trailing_zeroes_with_only_trailing_zeroes_keeps_the_decimal_point_by_default()
+{
+	std::string res = remove_trailing_zeroes("100.000");
+	ASSERT_EQUAL("100.0", res)
+}
+
+void
+test_left_trim_does_remove_all_leading_whitespaces()
+{
+	ASSERT_EQUAL("Something", left_trim(" \t\v\n\rSomething"))
+}
+
+void
+test_left_trim_does_not_remove_any_trailing_whitespaces()
+{
+	ASSERT_EQUAL("Something \t\v\n\r", left_trim("Something \t\v\n\r"))
+}
+
+void
+test_left_trim_does_not_remove_any_intermediate_whitespaces()
+{
+	ASSERT_EQUAL("Some \t\v\n\rthing", left_trim("Some \t\v\n\rthing"))
+}
+
+void
+test_left_trim_keeps_string_unmodified_if_no_leading_whitespace()
+{
+	ASSERT_EQUAL("Something", left_trim("Something"))
+}
+
+void
+test_right_trim_does_remove_all_trailing_whitespaces()
+{
+	ASSERT_EQUAL("Something", right_trim("Something \t\v\n\r"))
+}
+
+void
+test_right_trim_does_not_remove_any_leading_whitespaces()
+{
+	ASSERT_EQUAL(" \t\v\n\rSomething", right_trim(" \t\v\n\rSomething"))
+}
+
+void
+test_right_trim_does_not_remove_any_intermediate_whitespaces()
+{
+	ASSERT_EQUAL("Some \t\v\n\rthing", right_trim("Some \t\v\n\rthing"))
+}
+
+void
+test_right_trim_keeps_string_unmodified_if_no_trailing_whitespace()
+{
+	ASSERT_EQUAL("Something", right_trim("Something"))
+}
+
+void
+test_trim_does_remove_all_leading_or_trailing_whitespaces()
+{
+	ASSERT_EQUAL("Something", trim(" \t\v\n\rSomething \t\v\n\r"))
+}
+
+void
+test_trim_does_not_remove_any_intermediate_whitespaces()
+{
+	ASSERT_EQUAL("Some \t\v\n\rthing", trim("Some \t\v\n\rthing"))
+}
+
+void
+test_trim_keeps_string_unmodified_if_no_leading_or_trailing_whitespace()
+{
+	ASSERT_EQUAL("Something", trim("Something"))
+}
+
+/* === E N T R Y P O I N T ================================================= */
+
+int main() {
+
+	char* orig_LANG = getenv("LANG");
+	setenv("LANG", "en", 1);
```

Comment:
If this is used for number formatting, perhaps it's better to use `std::setlocale(...)`?

Reply:
Yes, it is.
But for some reason I can't remember, we force setlocale in the method

Reply:
I think because of this comment:
> // MinGW C++ std::locale accepts "C" and "POSIX" it does not support other locales.

But in this case it's safe to use "C" locale for this test.

Reply:
Is this it what you suggest?

Reply:
If I understand correctly this should be something like:
`synfig::ChangeLocale change_locale(LC_NUMERIC, "C");`

Reply:
It won't work.
There is a ` synfig::ChangeLocale change_locale(LC_NUMERIC, "");` inside `synfig::remove_trailing_zeroes()`
https://github.com/synfig/synfig/blob/aa8f1f35bcd67a7aaf765b6c7b3e152a39cfc1af/synfig-core/src/synfig/string_helper.cpp#L42-L56

Reply:
I wonder if line 47 is still really needed (` 	synfig::ChangeLocale changeLocale(LC_NUMERIC, ""); `).

The non-MinGW32 'version' just get the current locale name (`setlocale(LC_NUMERIC, nullptr)`) and create an object with this locale name. It doesn't actually change the locale.

MinGW32 does not support this strategy, as you pointed out, but `localeconv()` already gets the current locale info. Is there need to force the locale?

Analysing via `git blame` and `git log`, the first usage of `synfig::remove_trailing_zeroes()` is with the commit fd1e5ecdc3cf2cf9cb2ee642ca7f7b89aa04b40b (PR #2102), although the method was introduced (but not used anywhere) in a509ebd64fd7af777130abd9753e2e333ffb8575 (PR #1995). At that time, it didn't force a locale.

Later, to fix the MingW issue, we introduced the ChangeLocale in 3c0f04d55e54459e36b7872812a07772989e74d0 (PR #2157). Maybe we missed it there?

Reply:
I am not sure, but maybe this line is needed to read locale set by environment variables. Without it it should always return "C" locale.

https://man7.org/linux/man-pages/man3/setlocale.3.html
> If locale is an empty string, "", each part of the locale that
> should be modified is set according to the environment variables.

Also related discussion: 
https://stackoverflow.com/questions/33117258/c-setlocale-not-working-as-expect

Reply:
> If locale is a null pointer, setlocale queries the current C locale without modifying it. 
(https://en.cppreference.com/w/c/locale/setlocale)



> Also related discussion:
> https://stackoverflow.com/questions/33117258/c-setlocale-not-working-as-expect

This is about user setting wrong locale name (wrong: `en_US`, correct: `en_US.utf8`)

Reply:
> > Also related discussion:
> > https://stackoverflow.com/questions/33117258/c-setlocale-not-working-as-expect
> 
> This is about user setting wrong locale name (wrong: `en_US`, correct: `en_US.utf8`)

He asks two questions. Second was:
> When try to get current LC_TIME, it just print C, but not the value specified from environment variable.

And the comments points the need to read the values from the environment first.

Reply:
Here are where we already set locale:

https://github.com/synfig/synfig/blob/b836a72d4149552d562863db62efd5fa567fbe51/synfig-core/src/tool/main.cpp#L78-L82

https://github.com/synfig/synfig/blob/b836a72d4149552d562863db62efd5fa567fbe51/synfig-studio/src/gui/main.cpp#L61-L71

Reply:
It's still fails with `de_DE` locale (`LC_NUMERIC=de_DE.UTF-8 ./string`). 
Can you check the issue?

Reply:
What about now?
It's necessary to test on MS Windows...

Reply:
It works on Linux. Test on Windows today.

Reply:
> It works on Linux. Test on Windows today.

Could you test it?

Reply:
Sorry, been busy testing vcpkg.
Works without any issues.